src/common/snippets/src/pass/set_buffer_offset.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <snippets/itt.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+
+#include "snippets/snippets_isa.hpp"
+#include "snippets/pass/set_buffer_offset.hpp"
+#include "snippets/op/subgraph.hpp"
+
+
+ngraph::snippets::pass::SetBufferOffset::SetBufferOffset() {
+    MATCHER_SCOPE(SetBufferOffset);
+
+    auto m_buffer = ngraph::pattern::wrap_type<op::Buffer>();
+
+    auto callback = [&](ngraph::pattern::Matcher &m) {
+        OV_ITT_SCOPED_TASK(ngraph::pass::itt::domains::SnippetsTransform, "Snippets::op::SetBufferOffset")
+        auto root = m.get_match_root();
+        const auto buffer = ov::as_type_ptr<op::Buffer>(root);
+        buffer->set_offset(current_offset);
+
+        // If Buffer has offset We set this offset in the next Load and Store ops
+        // to correctly read and write data because all buffers have the one register
+        // Also if user sets offset to a Buffer It means that the Buffer has the corresponding Load and Store ops
+
+        // Propagate to up: in Store. Buffer can have only one Store
+        {
+            auto parent = buffer->get_input_node_shared_ptr(0);
+            auto idx = buffer->input(0).get_source_output().get_index();
+            while (std::dynamic_pointer_cast<snippets::op::LoopBase>(parent)) {
+                const auto source_output = parent->input_value(idx);
+                parent = source_output.get_node_shared_ptr();
+                idx = source_output.get_index();
+            }
+            if (auto store = std::dynamic_pointer_cast<snippets::op::Store>(parent)) {
+                store->set_offset(current_offset);
+            } else if (const auto brgemm = std::dynamic_pointer_cast<snippets::op::Brgemm>(parent)) {
+                // Brgemm encapsulates work with loading and storing of data
+                brgemm->set_offset_c(current_offset);
+            } else {
+                throw ngraph_error("Buffer::set_offset() was called when Buffer didn't have the corresponding Store op for offset propagation");
+            }
+        }
+
+        // Propagate to down: in Load. Buffer can have several Load and Loops after himself. We should go through all target inputs
+        {
+            std::function<void(const Input<Node>&)> propagate_down;
+            propagate_down = [&](const Input<Node>& target_input) {
```

Comment:
Thank you for the detailed example! But honestly I don't understand the possible problems. The pass matches on `Buffer` node and updates the corresponding operations like `Store`, `Load`, `Brgemm`. To find them we should go over all `LoopBase` ops. I mean that `Buffer` node has own `MemoryAccess` nodes and they're different for each `Buffer` node. Yeah, `Buffer` nodes can have the same `LoopBase` nodes but we don't touch them, only `MemoryAccess`

Reply:
Ok, then subsequent matcher-pass that modifies `MemoryAccess` can overwrite  your changes (it can be completely different transformation not connected to the `Buffer`)

Reply:
As we discussed offline, I left the comment
[8967b35](https://github.com/a-sidorova/openvino/pull/57/commits/8967b353a1f77e3d401d14c059e0d02429d791f0)